src/engine/Union.cpp

```diff
@@ -233,4 +237,73 @@ void Union::computeUnion(
     writeColumn(left, targetColumn, leftCol, 0u);
     writeColumn(right, targetColumn, rightCol, left.size());
   }
+  return res;
+}
+
+// _____________________________________________________________________________
+template <bool left>
+std::vector<size_t> Union::computePermutation() const {
+  size_t startOfUndefColumns = _subtrees[left ? 0 : 1]->getResultWidth();
+  std::vector<size_t> permutation(_columnOrigins.size());
+  for (size_t targetColIdx = 0; targetColIdx < _columnOrigins.size();
+       ++targetColIdx) {
+    size_t originIndex = _columnOrigins.at(targetColIdx)[left ? 0 : 1];
+    if (originIndex == NO_COLUMN) {
+      originIndex = startOfUndefColumns;
+      startOfUndefColumns++;
+    }
+    permutation[targetColIdx] = originIndex;
+  }
+  return permutation;
+}
+
+// _____________________________________________________________________________
+IdTable Union::transformToCorrectColumnFormat(
+    IdTable idTable, const std::vector<size_t>& permutation) const {
+  while (idTable.numColumns() < getResultWidth()) {
+    idTable.addEmptyColumn();
+    std::ranges::fill(idTable.getColumn(idTable.numColumns() - 1),
+                      Id::makeUndefined());
+  }
+
+  // i + 1 because once everything expect the last column is in the correct
+  // position, the last column is already in the correct position so the last
+  // iteration would always swap the last column with itself.
+  for (size_t i = 0; i + 1 < permutation.size(); ++i) {
+    size_t ind = permutation[i];
+    while (ind < i) {
+      ind = permutation[ind];
+    }
+    idTable.swapColumns(i, ind);
+  }
+  return idTable;
+}
+
+// _____________________________________________________________________________
+cppcoro::generator<IdTable> Union::computeResultLazily(
+    std::shared_ptr<const Result> result1,
+    std::shared_ptr<const Result> result2,
+    std::shared_ptr<LocalVocab> localVocab) const {
+  if (result1->isFullyMaterialized()) {
+    co_yield computeUnion(result1->idTable(),
+                          IdTable{getResultWidth(), allocator()},
+                          _columnOrigins);
+  } else {
+    std::vector<size_t> permutation = computePermutation<true>();
```

Comment:
I'm almost certain that for the left table this permutation will always be the identity function, so it might only be required to add columns for undef in this case